### PR TITLE
(WIP) Pulsar SQL supports pulsar's primitive schemas

### PR DESCRIPTION
**Pull request by sijie**
*Changes*

There is no *struct* in pulsar's primitive schemas. So introduced an internal fields `__value__`
for representing pulsar values.

Refactor following places:

- use `SchemaInfo` everywhere possible. We shouldn't strip out the information from schema info.
- avoid using avro schema directly.
- use pulsar `Schema` for serializing and deserializing and avoid reimplementing the same logic again and again, which can be problematic.
  also it allows taking all the changes we made to pulsar schema

*Tests*

WIP

*NOT IN SCOPE*

We should avoid handling avro schema or json schema here directly, because it doesn't handle schema versioning properly.
The right approach is to use Pulsar's GenericSchema which can handle schema versioning properly. But that refactor should
be done in a separated pull request.

### CODE REVIEW

**Comment by sijie** [comment]
@srkukarni 

Current presto integration bypasses Pulsar's schema's implementation and interacts with *avro schema data* directly. It introduces several problems:

1) all the *properties* are lost when presto gets splits.
2) it doesn't handle schema versions properly. 
3) it introduces a lot of duplicated code that has already been implemented in pulsar schema. it makes maintenance very hard.

Hence I don't think we should continue using the current approach for supporting other schema types. 

This is a WIP for demonstrating the approach I mentioned in #4022 to supporting primitive types in Presto.

File: pulsar-sql/presto-pulsar/src/main/java/org/apache/pulsar/sql/presto/PulsarSplit.java [hunk 1]
```diff
@@ -57,8 +56,7 @@ public PulsarSplit(
             @JsonProperty("schemaName") String schemaName,
             @JsonProperty("tableName") String tableName,
             @JsonProperty("splitSize") long splitSize,
-            @JsonProperty("schema") String schema,
-            @JsonProperty("schemaType") SchemaType schemaType,
+            @JsonProperty("schema") SchemaInfo schemaInfo,
```
Comment (jerrypeng) [inline pulsar-sql/presto-pulsar/src/main/java/org/apache/pulsar/sql/presto/PulsarSplit.java]: @sijie is this change necessary and are you sure it json serialization works here for SchemaInfo?  SchemaInfo info is a protobuf and JSON serializers don't work well with protobuf objects

Comment (sijie) [reply pulsar-sql/presto-pulsar/src/main/java/org/apache/pulsar/sql/presto/PulsarSplit.java]: SchemaInfo is a POJO not a protobuf generated class.

Comment (sijie) [reply pulsar-sql/presto-pulsar/src/main/java/org/apache/pulsar/sql/presto/PulsarSplit.java]: The reason I use `SchemaInfo` is we need all the information for constructing the right schema. A schema info is comprised of type, schema data and properties.  The presto implementation ignores the `properties` field which can potentially create bugs in future. So the change to use `SchemaInfo` is needed.

Comment (jerrypeng) [reply pulsar-sql/presto-pulsar/src/main/java/org/apache/pulsar/sql/presto/PulsarSplit.java]: my concern is that one of the fields ```byte[] schema```. JSON don't usually contain bytes and there could be issues with character encoding that is why I passed the schema in as a String in the first place.

Comment (jerrypeng) [reply pulsar-sql/presto-pulsar/src/main/java/org/apache/pulsar/sql/presto/PulsarSplit.java]: > The presto implementation ignores the properties field which can potentially create bugs in future.
can you elaborate?

Comment (sijie) [reply pulsar-sql/presto-pulsar/src/main/java/org/apache/pulsar/sql/presto/PulsarSplit.java]: If Presto is using Jackson (from the class, it appears to be using Jackson), Jackson handles `byte[]` properly as it use Base64 for encoding the bytes.

If you have concerns about this, I can change it to pass both `SchemaInfo` and `byte[]` and use Base64 for encoding. The current approach doesn't handle string encoding properly.

Comment (jerrypeng) [reply pulsar-sql/presto-pulsar/src/main/java/org/apache/pulsar/sql/presto/PulsarSplit.java]: I am fine with passing in SchemaInfo if all the encoding works properly.  For debugging purposes, can you adjust the toString() method so that the binary field ```schema``` in SchemaInfo gets printed out as a String.

Comment (sijie) [reply pulsar-sql/presto-pulsar/src/main/java/org/apache/pulsar/sql/presto/PulsarSplit.java]: > can you elaborate?

I mean current implementation only stores type and schema data in the split. It drops properties. This would cause the presto implementation ignores the metadata we attached to properties for a schema, which can create bugs. so whether we flatten all the fields from schema info or use schema info directly doesn't matter here. the point is we need all the fields in schema info.

**Comment by jerrypeng** [comment]
@sijie any additional progress on this?

**Comment by sijie** [comment]
@jerrypeng sorry I didn't make any more progress than current PR yet.

**Comment by jerrypeng** [comment]
@sijie what else are you planning on including in this PR?

**Comment by sijie** [comment]
@jerrypeng write the unit tests and integration tests. and also addressed your comment.

**Comment by sijie** [comment]
close this PR since @congbobo184 took over the task and continued the effort in #4692
